2020/02/28/consolidate-multiple-sftp-accounts-on-one-master-account.md

```diff
@@ -0,0 +1,153 @@
+
+
+Both the SFTP accounts as well as master account should be attached to one group. This will permit individual accounts to write on the master account-mounted path. In turn, the master account can read files from the same location. This location now has administrative rights of both the SFTP user as well as shared group permission. Group permission of the mounted folder is set to sftpgroup and user permission is set to facility account.
+
+
+A script is written to automate the list of actions in order to create an SFTP account, mount it at the master account path, and add fstab entries to save the mount in the case of a reboot. The script not only saved much time, but also avoided human error when creating accounts for all facilities in addition to future usage.
+
+* Create master account, sftpgroup and mount paths
+
+```bash
+
+# create master user account
+$ adduser master
+$ passwd master
+
+# Add master user to sftpgroup group
+usermod -a -G sftpgroup master
+getent group sftpgroup
+
+
+# MOUNT_PATH and sub folders
+mkdir -p /home/master/MOUNT_PATH/{Input,Pickup,Backup,Archive}
+chown -R master:master /home/master/MOUNT_PATH
+```
+
+* Automated script to create sftp account and mount folders
+
+```bash
+#!/bin/bash
+
+# sh create_sftp_account.sh sftpfacilityone FACILITY_ONE
+
+sftpuser=$1
+facility_name=$2
+
+# Create SFTP account and add to sftpgroup 
+useradd $sftpuser
+usermod -g sftpusers -s /usr/sbin/nologin $sftpuser
+usermod -a -G sftpgroup $sftpuser
+mkdir -p /home/$sftpuser
+chown root:root /home/$sftpuser
+mkdir -p /home/$sftpuser/INPUT_PATH
+
+# Create path specific to facility in master account Input folder
+mkdir -p /home/master/MOUNT_PATH/Input/$facility_name
+chown -R $sftpuser:sftpgroup /home/master/MOUNT_PATH/Input/$facility_name
+
+
+# Mount sftp account into master account and set permissions
+mount --bind /home/master/MOUNT_PATH/Input/$facility_name /home/$sftpuser/INPUT_PATH
+chown -R $sftpuser:sftpgroup /home/$sftpuser/INPUT_PATH
+chmod g=rwxs /home/$sftpuser/INPUT_PATH
+
+# Add fstab entry to persist and mount on reboot
+echo "/home/master/MOUNT_PATH/Input/$facility_name        /home/$sftpuser/INPUT_PATH        none        bind        0        0" >> /etc/fstab
+echo "Created user $sftpuser at $facility_name mount point successfully"
+```
+
+### Files to One Location
+
+Now, data files from facilities are available at individual folders under MOUNT_PATH/Input on the master account. These files need to be collected and moved to a single path where the application picks up and copies. The files are then stored in both Backup and Archive paths as processed files will be modified by the application. The following file script is scheduled at cron to run every minute and processes new files as they emerge:
```

Comment:
Just curious, if you need a script to consolidate all the contributed files into one directory, why do you need the bind mounts? Couldn't it read from the various user home directories, and move them to the destination?

Reply:
@joshwilliams Thats sounds interesting idea. I don't have any specific points to defend this bind mount method other than it is just another method implemented on the system.

What would be best choice to go with the blog post?
Option A - Remove this files consolidation by scripts section and justify to the title with mount at master account
Option B - Skip bind mount in account creation process and write a script to collect files from each accounts as per your idea
Option C - any other suggestions

On KDHE System implementation: Would you strongly recommend to use script to consolidate files instead of bind mount on security or best practice way ? If yes, i will check with Steve about modifying the implementation.

Reply:
That's fair -- if it were me, I'd just remove the section about using a script to move the files around. That way there's no question about other methods, plus it'll simplify and better focus the post. The rest of it boils down to the technique of using bind mounts to share files between separate chroot's, and I think that's a good thing to post about!

As far as our actual implementation, I'd look at just using the script. The bind mounts sound a little fragile, and even if not, it'd be good to simplify the configuration. But it doesn't sound like a high priority.

Thanks!

Reply:
I have removed scripts section in the blog post.

@joshwilliams  Thanks for wonderful feedback. I will watch out for opportunity to replace the bind mount method with script in future.